README
Book module for Moodle

CHANGES
CHANGES
=======

This file contains all the raw change logs from the past month,
so that you can see exactly what changes have been made recently.

---------------------------------------------------------------

2007-03-23 Friday 11:39  moodler

	* install/lang/: eu_utf8/installer.php, zh_cn_utf8/installer.php: 
	
	Automatic installer.php lang files by installer_builder (20070323)

2007-03-23 Friday 05:31  tjhunt

	* mod/quiz/report/grading/report.php: 
	
	MDL-7241 - Manual grading report, sort by completed time didn't.

2007-03-22 Thursday 11:53  moodler

	* install/lang/: el_utf8/installer.php, zh_cn_utf8/installer.php: 
	
	Automatic installer.php lang files by installer_builder (20070322)

2007-03-22 Thursday 07:25  tjhunt

	* mod/quiz/report/analysis/report.php: 
	
	MDL-7259 - Item analysis report does not use the right format_text
	settings.

2007-03-22 Thursday 07:04  tjhunt

	* mod/quiz/report/analysis/report.php: 
	
	Fix whitespace problems and Eclipse warnings in Item Analysis quiz
	report.

2007-03-22 Thursday 06:01  skodak

	* lib/moodlelib.php: 
	
	MDL-7647 Broken links on confirmation email (or others) - patch by
	Paulo Matos; backported from MOODLE_17_STABLE

2007-03-21 Wednesday 11:48  moodler

	* install/lang/: ta_utf8/installer.php, zh_cn_utf8/installer.php: 
	
	Automatic installer.php lang files by installer_builder (20070321)

2007-03-21 Wednesday 08:13  tjhunt

	* question/type/multianswer/questiontype.php: 
	
	MDL-8978 - Regression from MDL-6386.

2007-03-21 Wednesday 03:49  skodak

	* mod/resource/type/repository/hive/makelink.php: 
	
	MDL-8975 fix parameter handling in repository plugin; backported
	from MOODLE_18_STABLE

2007-03-21 Wednesday 00:25  skodak

	* admin/index.php: 
	
	SC#374 removing obsoleted utf8migration code; merged from
	MOODLE_17_STABLE

2007-03-20 Tuesday 22:31  tjhunt

	* question/type/: questiontype.php, multianswer/questiontype.php: 
	
	MDL-6386/MDL-6462 - yet more special character issues with
	multianswer. I think I have got them all this time. Backported from
	HEAD.

2007-03-20 Tuesday 11:33  moodler

	* install/lang/es_utf8/installer.php: 
	
	Automatic installer.php lang files by installer_builder (20070320)

2007-03-19 Monday 23:55  stronk7

	* filter/activitynames/filter.php: 
	
	small typo in var name
	
	Merged from MOODLE_15_STABLE

2007-03-19 Monday 14:03  martinlanghoff

	* filter/activitynames/filter.php, mod/glossary/filter.php,
	mod/resource/filter.php, mod/wiki/filter.php: 
	
	Auto-linking filters: fix serious caching bug in forum mailouts
	
	Autolinking of glossaries, activity, resources and wiki names were
	using a trivial single-entry cache implemented with static vars.
	The cache was _not_ keyed on course.
	
	This bug was visible during forum_cron() which walks many courses.
	The cache would get "stuck" on the first course that had something
	to put in the cache.  All mailouts from there onwards would
	autolink to stuff in the wrong course.

2007-03-18 Sunday 11:26  moodler

	* install/lang/sk_utf8/installer.php: 
	
	Automatic installer.php lang files by installer_builder (20070318)

2007-03-18 Sunday 06:58  poltawski

	* question/type/multianswer/questiontype.php: 
	
	MDL-6386 - Regression from MDL-6462, lost stripslashes in CLOZE
	question, thanks to Joseph Joseph RÃÂ©zeau  for his attention
	to this issue.	Backported from MOODLE_18_STABLE

2007-03-17 Saturday 06:29  pichetp

	* mod/quiz/view.php: 
	
	adding </p> lines 106,108

2007-03-15 Thursday 20:58  tjhunt

	* question/type/multianswer/questiontype.php: 
	
	MDL-6462 - Cloze (multianswer) question special character issues.

2007-03-15 Thursday 11:29  moodler

	* install/lang/de_utf8/installer.php: 
	
	Automatic installer.php lang files by installer_builder (20070315)

2007-03-15 Thursday 05:23  tjhunt

	* mod/quiz/db/: postgres7.php, mysql.php: 
	
	MDL-8127 - bad use of DESCRIPTION constant in quiz module causes
	upgrade to fail. Thanks to Pavel Krejci for the fix.

2007-03-14 Wednesday 23:15  skodak

	* filter/: algebra/algebradebug.php, algebra/filter.php,
	algebra/pix.php, tex/texed.php, tex/pix.php, tex/texdebug.php: 
	
	MDL-8897 fixed shell commands in tex and algebra filters;
	backported from MOODLE_17_STABLE

2007-03-14 Wednesday 21:48  ethem

	* enrol/authorize/authorizenetlib.php: 
	
	Code cleanup

2007-03-14 Wednesday 11:48  moodler

	* install/lang/eu_utf8/installer.php: 
	
	Automatic installer.php lang files by installer_builder (20070314)

2007-03-14 Wednesday 01:43  tjhunt

	* backup/restorelib.php: 
	
	Eliminate bizarre use of function_exists.

2007-03-13 Tuesday 11:31  moodler

	* install/lang/pt_utf8/installer.php: 
	
	Automatic installer.php lang files by installer_builder (20070313)

2007-03-12 Monday 16:54  skodak

	* lib/phpmailer/class.phpmailer.php: 
	
	MDL-5229 and MDL-6196 fixed infinite loop for gb18030 encoding,
	line breaking for gb18030 does not work though

2007-03-12 Monday 04:01  skodak

	* blocks/login/block_login.php: 
	
	MDL-8849 use proper value quoting; backported from MOODLE_18_STABLE

2007-03-10 Saturday 22:52  moodler

	* lib/timezones.txt: 
	
	Merged from head   MDL-8842

2007-03-10 Saturday 11:26  moodler

	* install/lang/da_utf8/installer.php: 
	
	Automatic installer.php lang files by installer_builder (20070310)

2007-03-10 Saturday 08:34  tjhunt

	* question/type/random/questiontype.php: 
	
	MDL-8177 - Random questions selected deleted questions. Backported
	from MOODLE_18_STABLE.

2007-03-09 Friday 11:43  moodler

	* install/lang/fr_utf8/installer.php: 
	
	Automatic installer.php lang files by installer_builder (20070309)

2007-03-08 Thursday 11:28  moodler

	* lib/statslib.php: 
	
	chaning require_once to include_once to tolerate faulty mods

2007-03-07 Wednesday 02:00  tjhunt

	* question/type/multichoice/questiontype.php: 
	
	MDL-8349 - responses not being filtered in history display.
	Backported from MOODLE_18_STABLE.

2007-03-05 Monday 20:30  moodler

	* lib/moodlelib.php: 
	
	Mrged new string locations for modules


view.php
<?PHP // $Id: view.php,v 1.2 2006/10/16 08:05:26 gustav_delius Exp $

require_once('../../config.php');
require_once('lib.php');

$id        = required_param('id', PARAM_INT);           // Course Module ID
$chapterid = optional_param('chapterid', 0, PARAM_INT); // Chapter ID
$edit      = optional_param('edit', -1, PARAM_BOOL);     // Edit mode

// =========================================================================
// security checks START - teachers edit; students view
// =========================================================================
if ($CFG->forcelogin) {
    require_login();
}

if (!$cm = get_record('course_modules', 'id', $id)) {
    error('Course Module ID is incorrect');
}

if (!$course = get_record('course', 'id', $cm->course)) {
    error('Course is misconfigured');
}

if ($course->category) {
    require_login($course->id);
}

if (!$book = get_record('book', 'id', $cm->instance)) {
    error('Course module is incorrect');
}

$isteacher = isteacheredit($course->id);
if ($isteacher) {
    if($edit != -1) {
        $USER->editing = $edit;
    } else {
        if (isset($USER->editing)) {
            $edit = $USER->editing;
        } else {
            $edit = 0;
        }
    }
} else {
    $edit = 0;
}

if (!$cm->visible and !$isteacher) {
    notice(get_string('activityiscurrentlyhidden'));
}

/// read chapters
$select = $isteacher ? "bookid = $book->id" : "bookid = $book->id AND hidden = 0";
$chapters = get_records_select('book_chapters', $select, 'pagenum', 'id, pagenum, subchapter, title, hidden');

if (!$chapters) {
    if ($isteacher) {
        redirect('edit.php?id='.$cm->id); //no chapters - add new one
        die;
    } else {
        error('Error reading book chapters.');
    }
}
/// check chapterid and read chapter data
if ($chapterid == '0') { // go to first chapter if no given
    foreach($chapters as $ch) {
        if ($isteacher) {
            $chapterid = $ch->id;
            break;
        }
        if (!$ch->hidden) {
            $chapterid = $ch->id;
            break;
        }
    }
}


if (!$chapter = get_record('book_chapters', 'id', $chapterid)) {
    error('Error reading book chapters.');
}

//check all variables
unset($id);
unset($chapterid);

/// chapter is hidden for students
if (!$isteacher and $chapter->hidden) {
    error('Error reading book chapters.');
}

/// chapter not part of this book!
if ($chapter->bookid != $book->id) {
    error('Chapter not part of this book!');
}
// =========================================================================
// security checks  END
// =========================================================================

add_to_log($course->id, 'book', 'view', 'view.php?id='.$cm->id.'&chapterid='.$chapter->id, $book->id, $cm->id);


///read standard strings
$strbooks = get_string('modulenameplural', 'book');
$strbook  = get_string('modulename', 'book');
$strTOC = get_string('TOC', 'book');

/// prepare header
if ($course->category) {
    $navigation = '<a href="../../course/view.php?id='.$course->id.'">'.$course->shortname.'</a> ->';
} else {
    $navigation = '';
}

$buttons = $isteacher ? '<table cellspacing="0" cellpadding="0"><tr><td>'.update_module_button($cm->id, $course->id, $strbook).'</td>'.
           '<td>&nbsp;</td><td>'.book_edit_button($cm->id, $course->id, $chapter->id).'</td></tr></table>'
           : '&nbsp;';

print_header( "$course->shortname: $book->name ($chapter->title)",
              $course->fullname,
              "$navigation <a href=\"index.php?id=$course->id\">$strbooks</a> -> $book->name",
              '',
              '<style type="text/css">@import url('.$CFG->wwwroot.'/mod/book/book_theme.css);</style>',
              true,
              $buttons,
              navmenu($course, $cm)
            );


/// prepare chapter navigation icons
$previd = null;
$nextid = null;
$found = 0;
foreach ($chapters as $ch) {
    if ($found) {
        $nextid= $ch->id;
        break;
    }
    if ($ch->id == $chapter->id) {
        $found = 1;
    }
    if (!$found) {
        $previd = $ch->id;
    }
}
if ($ch == current($chapters)) {
    $nextid = $ch->id;
}
$chnavigation = '';
if ($previd) {
    $chnavigation .= '<a title="'.get_string('navprev', 'book').'" href="view.php?id='.$cm->id.'&chapterid='.$previd.'"><img src="pix/nav_prev.gif" height="24" width="24" border="0"></img></a>';
} else {
    $chnavigation .= '<img src="pix/nav_prev_dis.gif" height="24" width="24" border="0"></img>';
}
$chnavigation .= '&nbsp;&nbsp;<img src="pix/nav_sep.gif" height="24" width="2" border="0"></img>&nbsp;&nbsp;';
if ($nextid) {
    $chnavigation .= '<a title="'.get_string('navnext', 'book').'" href="view.php?id='.$cm->id.'&chapterid='.$nextid.'"><img src="pix/nav_next.gif" height="24" width="24" border="0"></img></a>';
} else {
    $sec = '';
    if ($section = get_record('course_sections', 'id', $cm->section)) {
        $sec = $section->section;
    }
    $chnavigation .= '<a title="'.get_string('navexit', 'book').'" href="../../course/view.php?id='.$course->id.'#section-'.$sec.'"><img src="pix/nav_exit.gif" height="24" width="24" border="0"></img></a>';
}

/// prepare print icons
if ($book->disableprinting) {
    $printbook = '';
    $printchapter = '';
} else {
    $printbook = '<a title="'.get_string('printbook', 'book').'" href="print.php?id='.$cm->id.'" target="_blank"><img src="pix/print_book.gif" height="24" width="24" border="0"></img></a>';
    $printchapter = '<a title="'.get_string('printchapter', 'book').'" href="print.php?id='.$cm->id.'&chapterid='.$chapter->id.'" target="_blank"><img src="pix/print_chapter.gif" height="24" width="24" border="0"></img></a>';
}

// prepare $toc and $currtitle, $currsubtitle
require('toc.php');

if ($edit) {
    $tocwidth = $CFG->book_tocwidth + 80;
} else {
    $tocwidth = $CFG->book_tocwidth;
}

$doimport = ($isteacher and $edit) ? '<font size="-1"> (<a href="import.php?id='.$cm->id.'">'.get_string('doimport', 'book').'</a>)</font>' : '';


// =====================================================
// Book display HTML code
// =====================================================

?>
<table border="0" cellspacing="0" width="100%" valign="top" cellpadding="2">

<!-- subchapter title and upper navigation row //-->
<tr>
    <td width="<?php echo $tocwidth ?>" valign="bottom">
        <?php print_string('toc', 'book'); echo $doimport; ?>
    </td>
    <td valign="top">
        <table border="0" cellspacing="0" width="100%" valign="top" cellpadding="0">
        <tr>
            <td nowrap align="left"><?php echo $printbook.'&nbsp'.$printchapter ?></td>
            <td nowrap align="right"><?php echo $chnavigation ?></td>
        </tr>
        </table>
    </td>
</tr>

<!-- toc and chapter row //-->
<tr>
    <td width="<?php echo $tocwidth ?>" valign="top" align="left">
        <?php
        print_simple_box_start('middle', '100%');
        echo $toc;
        print_simple_box_end();
        if ($isteacher and $edit) {
            echo '<font size="1"><br />';
            helpbutton('faq', get_string('faq','book'), 'book', true, true);
            echo '</font>';
        }
        ?>
    </td>
    <td valign="top" align="right">
        <?php
        print_simple_box_start('middle', '100%');
        $content = '';
        if (!$book->customtitles) {
          if ($currsubtitle == '&nbsp;') {
              $content .= '<p class="book_chapter_title">'.$currtitle.'</p>';
          } else {
              $content .= '<p class="book_chapter_title">'.$currtitle.'<br />'.$currsubtitle.'</p>';
          }
        }
        $content .= $chapter->content;

        $nocleanoption = new object();
        $nocleanoption->noclean = true;
        echo '<div class="book_content">';
        echo format_text($content, FORMAT_HTML, $nocleanoption, $course->id);
        echo '</div>';
        print_simple_box_end();
        /// lower navigation
        echo '<p>'.$chnavigation.'<p>';
        ?>
    </td>
</tr>
</table>

<?php
print_footer($course);

?>
